### DISCUSSION TITLE
Yggdrasil is pretty fast (Rsyncing a 5GB test file through the network)

### DISCUSSION THREAD

**Discussion by Incognify** [yggdrasil-network/yggdrasil-go#868] [General]
I'm thinking of implementing Yggdrasil as part of secured off-site backup solution and wanted to test the speed when rsyncing large files through the network. I was pleasantly surprised.

So, this was a pretty basic test where I was pulling a 5GB test file from one server to another. The first test was between two servers, one in Finland and one in the Netherlands. The test was ran over Yggdrasil and over IPv4 (The Finnish server doesn't have native IPv6 capability).

The second test was between two servers, one in Dallas, Texas, USA and the other one being the same Netherlands server as above. With native IPv6 in both locations, I was able to test that as well as Yggdrasil and IPv4.

Anyway, the results:

**Yggdrasil - Netherlands pulling file from Finland**
```
root@ygg-test:~# rsync --progress -avz -e "ssh -p 8305" root@[200:23af:xxxx:xxxx:xxxx:xx:xxxx:xxxx]:/root/yggspeed.test /root/yggspeed.test
receiving incremental file list
yggspeed.test
  5,368,709,120 100%  126.39MB/s    0:00:40 (xfr#1, to-chk=0/1)

sent 43 bytes  received 5,221,101 bytes  125,810.70 bytes/sec
total size is 5,368,709,120  speedup is 1,028.26

```

Above is pulling the 5GB test file over the Yggdrasil network with speeds of 126.39 MB/s. Not bad! Especially when compared to the clearnet/IPv4 speeds shown below:

**IPv4 - Netherlands pulling file from Finland**
```
root@ygg-test:~# rsync --progress -avz -e "ssh -p 8305" root@xx.xx.xx.xx:/root/yggspeed.test /root/yggspeed.test-v4
receiving incremental file list
yggspeed.test
  5,368,709,120 100%  129.91MB/s    0:00:39 (xfr#1, to-chk=0/1)

sent 43 bytes  received 5,221,101 bytes  120,026.30 bytes/sec
total size is 5,368,709,120  speedup is 1,028.26
```

So, as you can see, the clearnet speed was barely faster, coming in at 129.91MB/s, a difference of 3.52MB/s.

This isn't a very scientific test. Geographically, both these servers are relatively close to one another (Finland and Netherlands) so that may have something to do with the performance being similar.

---

### Texas to Netherlands

**Yggdrasil - Netherlands pulling file from Texas, USA**

```
root@ygg-test:~# rsync --progress -avz -e "ssh -p 8305" root@[203:51cb:xxx:xxx:xxxx:xxxx:xxxx:xxxx]:/root/tx-to-nl.ygg.test /root/tx-to-nl.ygg.test
receiving incremental file list
tx-to-nl.ygg.test
  5,368,709,120 100%  137.33MB/s    0:00:37 (xfr#1, to-chk=0/1)

sent 43 bytes  received 5,221,106 bytes  128,917.26 bytes/sec
total size is 5,368,709,120  speedup is 1,028.26
```

At 137.33MB/s, we're faster than we were moving the same 5GB test file between the Netherlands and Finland. Interesting.

Let's try moving the same file over traditional IPv6, which is usually noticeably faster than IPv4 over longer routes:

**IPv6 - Netherlands pulling file from Texas, USA**

```
root@ygg-test:~# rsync --progress -avz -e "ssh -p 8305" root@[2604:fbc0:x:xx:xxx:xx:xxx:xxx]:/root/tx-to-nl.ygg.test /root/tx-to-nl.ygg.test-v6
receiving incremental file list
tx-to-nl.ygg.test
  5,368,709,120 100%  200.56MB/s    0:00:25 (xfr#1, to-chk=0/1)

sent 43 bytes  received 5,221,106 bytes  171,185.21 bytes/sec
total size is 5,368,709,120  speedup is 1,028.26
```

Wow, so that is a _big difference_ in speed between Yggdrasil's v6 and the clearnet v6.

Let's see how clearnet v6 compares to traditional IPv4:

**IPv4 - Netherlands pulling file from Texas, USA**

```
root@ygg-test:~# rsync --progress -avz -e "ssh -p 8305" root@[xxx.xxx.xxx.xxx]:/root/tx-to-nl.ygg.test /root/tx-to-nl.ygg.test-v4
receiving incremental file list
tx-to-nl.ygg.test
  5,368,709,120 100%  196.98MB/s    0:00:25 (xfr#1, to-chk=0/1)

sent 43 bytes  received 5,221,106 bytes  165,750.76 bytes/sec
total size is 5,368,709,120  speedup is 1,028.26
```

Okay, so still a pretty big difference but Yggdrasil is still incredibly fast.

---

### Pulling 5GB Test File from Finland to the Netherlands:
- IPv4:  129.91MB/s
- IPv6: (Unsupported on Finland server, so no test ran)
- Yggdrasil: 126.39MB/s

### Pulling 5GB Test File from Dallas, Texas to the Netherlands:
- IPv4:  196.98MB/s
- IPv6: 200.56MB/s
- Yggdrasil: 137.33MB/s

So yeah, it's not very scientific and running a single test for each location and network isn't really going to give me the most clear picture but at a glance, it shows that Yggdrasil is still pretty damn fast and capable.

**Comment by neilalexander** [comment]
There are no doubt some optimisations that can be made to improve performance further, like by reducing allocations in the code, but I think these figures are probably good enough for most applications at least. Particularly where dedicated peerings or fast routes are being used.

**Comment by 621625** [comment]
How does it scale? Let's say you get one million users, how do you ensure that it doesn't get clogged up?
